Rithika taking over for Aditya-feat: Add Material Cost Correlation API endpoint

Aditya-gam opened the pull request:
# Description

Implements a new Material Cost Correlation API endpoint that aggregates material usage data from `updateRecord` arrays and cost data from approved `purchaseRecord` entries to analyze consumption vs cost efficiency across projects.
<img width="530" height="705" alt="IssueDescription1" src="https://github.com/user-attachments/assets/2817a605-b696-40c7-8344-c68d677d6b67" />
<img width="624" height="661" alt="IssueDescription2" src="https://github.com/user-attachments/assets/9e7aa8fc-6161-4a2e-9057-504c315e038c" />
<img width="331" height="335" alt="ExpectedGraph" src="https://github.com/user-attachments/assets/b18e39ba-3fb7-4180-b775-8c90c2e30024" />

---

## Related PRS (if any):

This PR is the backend for the frontend [PR 4587](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/4587).

---

## Main changes explained:

### Created/Updated Files:

1. **src/controllers/bmdashboard/bmMaterialsController.js**
	- Added `bmGetMaterialCostCorrelation` method with helper functions for query parameter extraction, date handling, and error handling
	
2. **src/utilities/materialCostCorrelationHelpers.js**
	- MongoDB aggregation pipelines for usage and cost data
	- Name resolution functions (project/material names → ObjectIds)
	- Response building and data merging utilities
	
3. **src/utilities/materialCostCorrelationDateUtils.js**
	- Date parsing (supports YYYY-MM-DD, MM-DD-YYYY, MM/DD/YYYY, ISO 8601)
	- UTC date normalization (start of day for start dates, end of day for end dates)
	- Special handling: end date "today" returns current time minus 5 minutes
	
4. **src/utilities/queryParamParser.js**
	- Multi-select query parameter parser (handles single values, comma-separated, arrays)
	- Optional ObjectId validation
	
5. **src/routes/bmdashboard/bmMaterialsRouter.js**
	- Added `GET /materials/cost-correlation` route (positioned before `/materials/:projectId`)
	
6. **Test files** (5 test files)
	- Unit tests for utilities, controller, and router
---
### API Endpoint:

**`GET /api/bm/materials/cost-correlation`**

**Query Parameters:**
- `projectId` / `projectName` (optional, multi-select): Filter by project(s)
- `materialType` / `materialName` (optional, multi-select): Filter by material type(s)
- `startDate` (optional): YYYY-MM-DD, MM-DD-YYYY, MM/DD/YYYY, or ISO 8601 (defaults to earliest material record)
- `endDate` (optional): Same formats (defaults to current time minus 5 minutes)

**Response Structure:**
```json
{
	"meta": {
		"request": { "projectIds": [], "materialTypeIds": [], "startDateInput": "...", "endDateInput": "..." },
		"range": { "effectiveStart": "...", "effectiveEnd": "...", "defaultsApplied": {...} },
		"units": { "currency": "USD", "costScale": { "raw": 1, "k": 1000 } }
	},

	"data": [
		{
			"projectId": "...",
			"projectName": "Project A",
			"totals": { "quantityUsed": 2500, "totalCost": 5000, "totalCostK": 5, "costPerUnit": 2 },
			"byMaterialType": [
				{
					"materialTypeId": "...",
					"materialTypeName": "Cement",
					"unit": "tons",
					"quantityUsed": 2500,
					"totalCost": 5000,
					"totalCostK": 5,
					"costPerUnit": 2
				}
			]
		}
	]
}
```
---
**Key Implementation Details:**
- Usage data aggregated from `BuildingMaterial.updateRecord` arrays
- Cost data aggregated from `BuildingMaterial.purchaseRecord` arrays (status='Approved' only)
- Aggregations run in parallel
- Dates normalized to UTC (start: 00:00:00.000Z, end: 23:59:59.999Z)
- Name-based filtering resolves names to ObjectIds via database queries
- Error handling: 400 (Bad Request), 422 (Unprocessable Entity), 500 (Internal Server Error)

---

## How to test:

1. Checkout branch: `Aditya-feat/material-cost-correlation-chart`
2. Reinstall dependencies and clean cache using `rm -rf node_modules package-lock.json && npm cache clean --force`
3. Run `npm install` to install dependencies, then run the backend locally (`npm run dev`)
4. Use the owner/admin login for the authentication token.
5. Test endpoint:
```bash
GET /api/bm/materials/cost-correlation?projectId=<id>&startDate=2024-01-01&endDate=2024-12-31

GET /api/bm/materials/cost-correlation?projectName=Project%20A&materialName=Cement
```
4. Run tests: `npm test -- bmMaterialsController materialCostCorrelation queryParamParser bmMaterialsRouter`

---

## Screenshots or videos of changes:

1. Light Mode:
<img width="457" height="773" alt="LightModeChart" src="https://github.com/user-attachments/assets/849258e3-f3a7-4ec7-8a17-c8ef5bcfe188" />

3. Dark Mode:
<img width="472" height="799" alt="DarkModeChart" src="https://github.com/user-attachments/assets/989b29f7-c20b-4e4f-a50c-0dca9acae832" />

5. Test Video:

https://github.com/user-attachments/assets/decfe7f9-6fb2-48ae-bdce-bfa137015bf7

---

## Note:

- **Database Changes**: Added mock data so the endpoint returns some data.
- **Backward Compatibility**: Fully compatible - new endpoint only
- **Performance**: Parallel aggregations and batched lookups

---

Aditya-gam committed:
feat: add date parsing and normalization utilities for material cost correlation

Implement comprehensive date utilities supporting multiple input formats:
- parseDateInput: handles ISO, MM-DD-YYYY, MM/DD/YYYY formats and Date objects
- normalizeStartDate: normalizes to start of day in UTC
- normalizeEndDate: normalizes to end of day UTC with special 'today' handling (now-5min)
- isDateToday: helper to check if date matches today in UTC
- parseAndNormalizeDateRangeUTC: main orchestrator with default handling and validation

All functions use UTC timezone and throw structured error objects following existing patterns.
Replaces inline date handling patterns found in existing controllers.

Aditya-gam committed:
feat: add date parsing and normalization utilities for material cost correlation

Implement comprehensive date utilities supporting multiple input formats:
- parseDateInput: handles ISO, MM-DD-YYYY, MM/DD/YYYY formats and Date objects
- normalizeStartDate: normalizes to start of day in UTC
- normalizeEndDate: normalizes to end of day UTC with special 'today' handling (now-5min)
- isDateToday: helper to check if date matches today in UTC
- parseAndNormalizeDateRangeUTC: main orchestrator with default handling and validation

All functions use UTC timezone and throw structured error objects following existing patterns.
Replaces inline date handling patterns found in existing controllers.

Aditya-gam committed:
feat: add multi-select query parameter parser utility

Implement generic query parameter parser supporting multiple input formats:
- Handles undefined, single values, comma-separated strings, and arrays
- Optional ObjectId validation using mongoose.Types.ObjectId.isValid
- Filters empty strings and whitespace-only values
- Throws structured error objects consistent with date parsing utilities
- Collects all invalid ObjectIds before throwing (not on first invalid)

Reuses ObjectId validation pattern from existing controllers and comma-splitting pattern from materialCostController. Designed to be generic and reusable across endpoints.

Aditya-gam committed:
feat: add multi-select query parameter parser utility

Implement generic query parameter parser supporting multiple input formats:
- Handles undefined, single values, comma-separated strings, and arrays
- Optional ObjectId validation using mongoose.Types.ObjectId.isValid
- Filters empty strings and whitespace-only values
- Throws structured error objects consistent with date parsing utilities
- Collects all invalid ObjectIds before throwing (not on first invalid)

Reuses ObjectId validation pattern from existing controllers and comma-splitting pattern from materialCostController. Designed to be generic and reusable across endpoints.

Aditya-gam committed:
feat: add material aggregation helper utilities for cost correlation

Implement MongoDB aggregation helpers for material usage and cost calculations:
- getEarliestRelevantMaterialDate: finds earliest date from updateRecord/purchaseRecord with filters
- aggregateMaterialUsage: aggregates quantityUsed from updateRecord arrays by date range
- aggregateMaterialCost: aggregates totalCost from approved purchaseRecord arrays by date range
- buildBaseMatchForMaterials: shared helper to build base match conditions (prevents duplication)

Reuses aggregation patterns from materialCostController (unwind, match, group, project stages).
Only counts purchases with status='Approved' matching existing endpoint behavior.
All functions return safe defaults (null/empty arrays) on error, logging exceptions for debugging.

Aditya-gam committed:
feat: add material aggregation helper utilities for cost correlation

Implement MongoDB aggregation helpers for material usage and cost calculations:
- getEarliestRelevantMaterialDate: finds earliest date from updateRecord/purchaseRecord with filters
- aggregateMaterialUsage: aggregates quantityUsed from updateRecord arrays by date range
- aggregateMaterialCost: aggregates totalCost from approved purchaseRecord arrays by date range
- buildBaseMatchForMaterials: shared helper to build base match conditions (prevents duplication)

Reuses aggregation patterns from materialCostController (unwind, match, group, project stages).
Only counts purchases with status='Approved' matching existing endpoint behavior.
All functions return safe defaults (null/empty arrays) on error, logging exceptions for debugging.

Aditya-gam committed:
refactor: reduce parameter count in aggregation functions

Group related parameters into objects to comply with linter max-parameters rule:
- aggregateMaterialUsage: group projectIds/materialTypeIds into filters object, effectiveStart/effectiveEnd into dateRange object
- aggregateMaterialCost: same parameter grouping applied
- Reduces parameter count from 5 to 3 (BuildingMaterial, filters, dateRange)
- Functions destructure objects internally, preserving existing behavior

Aditya-gam committed:
refactor: reduce parameter count in aggregation functions

Group related parameters into objects to comply with linter max-parameters rule:
- aggregateMaterialUsage: group projectIds/materialTypeIds into filters object, effectiveStart/effectiveEnd into dateRange object
- aggregateMaterialCost: same parameter grouping applied
- Reduces parameter count from 5 to 3 (BuildingMaterial, filters, dateRange)
- Functions destructure objects internally, preserving existing behavior

Aditya-gam committed:
feat: add response builder utility for material cost correlation

Implement buildCostCorrelationResponse function to merge usage and cost data:
- Merges usage and cost data by composite key (projectId-materialTypeId)
- Performs parallel lookups for project names and material type names/units
- Computes derived values (totalCostK, costPerUnit) with division-by-zero handling
- Groups data by project with per-material-type breakdowns
- Handles projects with explicit selection but no data (empty arrays)
- Sorts projects alphabetically by name
- Builds comprehensive meta object with request, range, and units info
- Returns structured response with meta and data arrays

Includes helper functions to prevent duplication:
- calculateCostPerUnit: handles division-by-zero, returns null when quantityUsed is 0
- calculateTotalCostK: converts cost to thousands
- objectIdToString: safely converts ObjectIds to strings

All edge cases handled: missing lookups use fallbacks, empty results return valid structure, errors return safe defaults.

Aditya-gam committed:
feat: add response builder utility for material cost correlation

Implement buildCostCorrelationResponse function to merge usage and cost data:
- Merges usage and cost data by composite key (projectId-materialTypeId)
- Performs parallel lookups for project names and material type names/units
- Computes derived values (totalCostK, costPerUnit) with division-by-zero handling
- Groups data by project with per-material-type breakdowns
- Handles projects with explicit selection but no data (empty arrays)
- Sorts projects alphabetically by name
- Builds comprehensive meta object with request, range, and units info
- Returns structured response with meta and data arrays

Includes helper functions to prevent duplication:
- calculateCostPerUnit: handles division-by-zero, returns null when quantityUsed is 0
- calculateTotalCostK: converts cost to thousands
- objectIdToString: safely converts ObjectIds to strings

All edge cases handled: missing lookups use fallbacks, empty results return valid structure, errors return safe defaults.

Aditya-gam committed:
feat: implement material cost correlation controller method

Add bmGetMaterialCostCorrelation controller function that orchestrates the entire request flow:
- Parses and validates multi-select query parameters (projectId, materialType) using utility
- Computes default start date from earliest material record if not provided
- Parses and normalizes date range with UTC handling and validation
- Runs usage and cost aggregations in parallel for performance
- Builds structured response with merged data, lookups, and computed values
- Comprehensive error handling with appropriate HTTP status codes (400, 422, 500)
- Logs all errors with request context for debugging
- Returns structured JSON response with meta and data arrays

Follows existing controller patterns and reuses all utility functions to prevent code duplication.

Aditya-gam committed:
feat: implement material cost correlation controller method

Add bmGetMaterialCostCorrelation controller function that orchestrates the entire request flow:
- Parses and validates multi-select query parameters (projectId, materialType) using utility
- Computes default start date from earliest material record if not provided
- Parses and normalizes date range with UTC handling and validation
- Runs usage and cost aggregations in parallel for performance
- Builds structured response with merged data, lookups, and computed values
- Comprehensive error handling with appropriate HTTP status codes (400, 422, 500)
- Logs all errors with request context for debugging
- Returns structured JSON response with meta and data arrays

Follows existing controller patterns and reuses all utility functions to prevent code duplication.

Aditya-gam committed:
refactor: extract magic numbers and replace console.error in bmMaterialsController

Extract magic numbers to named constants for better maintainability:
- DECIMAL_PRECISION = 4 for quantity calculations (.toFixed)
- DAYS_IN_WEEK = 7 and DAYS_IN_TWO_WEEKS = 14 for date calculations

Replace console.error with logger.logException in bmGetMaterialSummaryByProject:
- Use standard logger utility with request context (method, path, params, query)
- Replace magic number 500 with HTTP_STATUS_INTERNAL_SERVER_ERROR constant
- Maintains consistency with other controller error handling patterns

Aditya-gam committed:
refactor: extract magic numbers and replace console.error in bmMaterialsController

Extract magic numbers to named constants for better maintainability:
- DECIMAL_PRECISION = 4 for quantity calculations (.toFixed)
- DAYS_IN_WEEK = 7 and DAYS_IN_TWO_WEEKS = 14 for date calculations

Replace console.error with logger.logException in bmGetMaterialSummaryByProject:
- Use standard logger utility with request context (method, path, params, query)
- Replace magic number 500 with HTTP_STATUS_INTERNAL_SERVER_ERROR constant
- Maintains consistency with other controller error handling patterns

Aditya-gam committed:
feat: add route for material cost correlation endpoint

Register GET /materials/cost-correlation route in bmMaterialsRouter:
- Route placed before /materials/:projectId to ensure proper Express route matching
- Full path: /api/bm/materials/cost-correlation
- Binds to controller.bmGetMaterialCostCorrelation method
- Follows existing route pattern and formatting style

Route order is critical - specific routes must come before parameterized routes to prevent Express from matching cost-correlation as a projectId parameter.

Aditya-gam committed:
feat: add route for material cost correlation endpoint

Register GET /materials/cost-correlation route in bmMaterialsRouter:
- Route placed before /materials/:projectId to ensure proper Express route matching
- Full path: /api/bm/materials/cost-correlation
- Binds to controller.bmGetMaterialCostCorrelation method
- Follows existing route pattern and formatting style

Route order is critical - specific routes must come before parameterized routes to prevent Express from matching cost-correlation as a projectId parameter.

Aditya-gam committed:
refactor: use HTTP status constant in error handling

Replace magic number 400 with HTTP_STATUS_BAD_REQUEST constant in date range error handling for consistency.

Error handling is already comprehensively implemented:
- Query parameter validation returns 400 for invalid ObjectIds
- Date parsing errors return 422 (Unprocessable Entity)
- Date range validation errors return 400 (Bad Request)
- Aggregation errors return 500 with safe defaults (empty arrays)
- Response building errors return 500 with fallback handling
- All errors are logged with request context using logger utility
- Division-by-zero handled gracefully (returns null)
- Missing lookups use fallback values (ID as name)
- Consistent error response format: { error: message }

Aditya-gam committed:
refactor: use HTTP status constant in error handling

Replace magic number 400 with HTTP_STATUS_BAD_REQUEST constant in date range error handling for consistency.

Error handling is already comprehensively implemented:
- Query parameter validation returns 400 for invalid ObjectIds
- Date parsing errors return 422 (Unprocessable Entity)
- Date range validation errors return 400 (Bad Request)
- Aggregation errors return 500 with safe defaults (empty arrays)
- Response building errors return 500 with fallback handling
- All errors are logged with request context using logger utility
- Division-by-zero handled gracefully (returns null)
- Missing lookups use fallback values (ID as name)
- Consistent error response format: { error: message }

Aditya-gam committed:
refactor: extract ObjectId conversion to eliminate duplication

Extract repeated ObjectId conversion pattern into convertStringsToObjectIds helper:
- Eliminates 4 instances of idStrings.map((id) => new mongoose.Types.ObjectId(id))
- Used in buildBaseMatchForMaterials and buildCostCorrelationResponse
- Reduces code duplication and improves maintainability

Add comprehensive duplication review document verifying:
- All utility functions are reusable and centralized
- No business logic duplication between utilities and controller
- Error handling follows consistent patterns
- All calculations use shared helpers
- Code duplication is below 3% target
- All checklist items from implementation plan verified

Aditya-gam committed:
refactor: extract ObjectId conversion to eliminate duplication

Extract repeated ObjectId conversion pattern into convertStringsToObjectIds helper:
- Eliminates 4 instances of idStrings.map((id) => new mongoose.Types.ObjectId(id))
- Used in buildBaseMatchForMaterials and buildCostCorrelationResponse
- Reduces code duplication and improves maintainability

Add comprehensive duplication review document verifying:
- All utility functions are reusable and centralized
- No business logic duplication between utilities and controller
- Error handling follows consistent patterns
- All calculations use shared helpers
- Code duplication is below 3% target
- All checklist items from implementation plan verified

Aditya-gam committed:
test: move queryParamParser test to __tests__ directory

Move test file from src/utilities/queryParamParser.test.js to src/utilities/__tests__/queryParamParser.test.js to follow repository convention:
- Matches pattern used by bmdashboard controllers (__tests__ subdirectories)
- Keeps test files organized and separate from source files
- Updates import path from './queryParamParser' to '../queryParamParser'

All 40 tests pass with 100% coverage. This convention will be followed for all future utility test files.

Aditya-gam committed:
test: move queryParamParser test to __tests__ directory

Move test file from src/utilities/queryParamParser.test.js to src/utilities/__tests__/queryParamParser.test.js to follow repository convention:
- Matches pattern used by bmdashboard controllers (__tests__ subdirectories)
- Keeps test files organized and separate from source files
- Updates import path from './queryParamParser' to '../queryParamParser'

All 40 tests pass with 100% coverage. This convention will be followed for all future utility test files.

Aditya-gam committed:
test: add comprehensive tests for materialCostCorrelationDateUtils

Implement 88 unit tests covering all date utility functions:
- parseDateInput: valid/invalid formats, error structure
- normalizeStartDate: UTC/local normalization, edge cases
- isDateToday: UTC/local comparison, boundary conditions
- normalizeEndDate: today capping, non-today normalization, edge cases
- parseAndNormalizeDateRangeUTC: date range validation, default handling, error cases

Achieves 94% statement coverage, 86% branch coverage, 100% function coverage.
All tests use fixed date mocking for consistent results. Tests follow repository
convention in __tests__ subdirectory.

Aditya-gam committed:
test: add comprehensive tests for materialCostCorrelationDateUtils

Implement 88 unit tests covering all date utility functions:
- parseDateInput: valid/invalid formats, error structure
- normalizeStartDate: UTC/local normalization, edge cases
- isDateToday: UTC/local comparison, boundary conditions
- normalizeEndDate: today capping, non-today normalization, edge cases
- parseAndNormalizeDateRangeUTC: date range validation, default handling, error cases

Achieves 94% statement coverage, 86% branch coverage, 100% function coverage.
All tests use fixed date mocking for consistent results. Tests follow repository
convention in __tests__ subdirectory.

Aditya-gam committed:
test: add comprehensive tests for materialCostCorrelationHelpers

Implement 86 unit tests covering all helper functions:
- convertStringsToObjectIds: array conversion, edge cases
- buildBaseMatchForMaterials: project/material filtering, combined filters
- getEarliestRelevantMaterialDate: parallel queries, result comparison, error handling
- aggregateMaterialUsage: pipeline structure, filtering, return format
- aggregateMaterialCost: status filtering, cost calculation, field validation
- calculateCostPerUnit: division handling, edge cases, rounding
- calculateTotalCostK: cost scaling calculations
- objectIdToString: conversion handling, null/undefined cases
- buildCostCorrelationResponse: lookup maps, data merging, project totals, sorting, meta construction, error handling

Exports helper functions for testing. Achieves 97% statement coverage, 76% branch coverage, 95% function coverage. All tests use mocked Mongoose models and logger.

Aditya-gam committed:
test: add comprehensive tests for materialCostCorrelationHelpers

Implement 86 unit tests covering all helper functions:
- convertStringsToObjectIds: array conversion, edge cases
- buildBaseMatchForMaterials: project/material filtering, combined filters
- getEarliestRelevantMaterialDate: parallel queries, result comparison, error handling
- aggregateMaterialUsage: pipeline structure, filtering, return format
- aggregateMaterialCost: status filtering, cost calculation, field validation
- calculateCostPerUnit: division handling, edge cases, rounding
- calculateTotalCostK: cost scaling calculations
- objectIdToString: conversion handling, null/undefined cases
- buildCostCorrelationResponse: lookup maps, data merging, project totals, sorting, meta construction, error handling

Exports helper functions for testing. Achieves 97% statement coverage, 76% branch coverage, 95% function coverage. All tests use mocked Mongoose models and logger.

Aditya-gam committed:
test: add comprehensive tests for bmGetMaterialCostCorrelation

Implement 35 unit tests covering all request flow scenarios:
- Successful flows: all parameters, default dates, no filters
- Query parameter validation: invalid ObjectIds, empty parameters
- Date parsing errors: invalid formats, invalid ranges
- Aggregation errors: usage/cost failures, database errors
- Response building errors: lookup failures, logic errors
- Default date computation: earliest date, fallbacks, error handling
- Parallel execution: Promise.all verification
- Response structure: meta/data validation
- Edge cases: empty results, missing properties
- Logging verification: all error paths logged with context

All tests use mocked utilities and models. Tests follow repository
convention and extend existing test file.

Aditya-gam committed:
test: add comprehensive tests for bmGetMaterialCostCorrelation

Implement 35 unit tests covering all request flow scenarios:
- Successful flows: all parameters, default dates, no filters
- Query parameter validation: invalid ObjectIds, empty parameters
- Date parsing errors: invalid formats, invalid ranges
- Aggregation errors: usage/cost failures, database errors
- Response building errors: lookup failures, logic errors
- Default date computation: earliest date, fallbacks, error handling
- Parallel execution: Promise.all verification
- Response structure: meta/data validation
- Edge cases: empty results, missing properties
- Logging verification: all error paths logged with context

All tests use mocked utilities and models. Tests follow repository
convention and extend existing test file.

Aditya-gam committed:
test: add router tests for bmMaterialsRouter cost-correlation route

Implement 10 unit tests covering route registration and structure:
- Route registration: correct path, HTTP method, controller binding
- Route ordering: cost-correlation before parameterized route
- Router structure: Express Router instance, all routes registered

Tests verify that /materials/cost-correlation route is registered
before /materials/:projectId to ensure correct route matching.
All tests use mocked controller and verify route structure.

Aditya-gam committed:
test: add router tests for bmMaterialsRouter cost-correlation route

Implement 10 unit tests covering route registration and structure:
- Route registration: correct path, HTTP method, controller binding
- Route ordering: cost-correlation before parameterized route
- Router structure: Express Router instance, all routes registered

Tests verify that /materials/cost-correlation route is registered
before /materials/:projectId to ensure correct route matching.
All tests use mocked controller and verify route structure.

Aditya-gam committed:
feat: add name-based query parameters for material cost correlation

Add support for projectName and materialName query parameters in addition to existing ID-based parameters. Implement name resolution functions to convert names to ObjectIds with proper error handling. Maintain backward compatibility with existing projectId and materialType parameters.

Aditya-gam committed:
feat: add name-based query parameters for material cost correlation

Add support for projectName and materialName query parameters in addition to existing ID-based parameters. Implement name resolution functions to convert names to ObjectIds with proper error handling. Maintain backward compatibility with existing projectId and materialType parameters.

Aditya-gam committed:
refactor: improve cost correlation data handling and reduce complexity

- Add explicit date existence checks in purchase record aggregation
- Extract helper functions to reduce buildCostCorrelationResponse complexity
- Add defensive null/undefined checks in data merging logic
- Ensure totalCost is properly converted to number type
- Improve error handling and data validation

Aditya-gam committed:
refactor: improve cost correlation data handling and reduce complexity

- Add explicit date existence checks in purchase record aggregation
- Extract helper functions to reduce buildCostCorrelationResponse complexity
- Add defensive null/undefined checks in data merging logic
- Ensure totalCost is properly converted to number type
- Improve error handling and data validation

Aditya-gam committed:
fix(bm): use correct BuildingMaterial model for cost correlation API

Fix critical bug where cost correlation endpoint was querying wrong MongoDB collection.

Changes:
- routes.js: Switch from buildingMaterial (buildingInventoryItems collection)
  to buildingMaterialModel (buildingMaterials collection) which contains
  unitPrice data
- materialCostCorrelationHelpers.js:
  - Add  stage to handle missing unitPrice values in aggregation
  - Improve cost value validation (NaN checks) in mergeUsageAndCostData
  - Add defensive validation in buildProjectsMap for data integrity
  - Enhance error handling with console.error logging
- bmMaterialsController.js: Add response validation check for debugging.

Aditya-gam committed:
fix(bm): use correct BuildingMaterial model for cost correlation API

Fix critical bug where cost correlation endpoint was querying wrong MongoDB collection.

Changes:
- routes.js: Switch from buildingMaterial (buildingInventoryItems collection)
  to buildingMaterialModel (buildingMaterials collection) which contains
  unitPrice data
- materialCostCorrelationHelpers.js:
  - Add  stage to handle missing unitPrice values in aggregation
  - Improve cost value validation (NaN checks) in mergeUsageAndCostData
  - Add defensive validation in buildProjectsMap for data integrity
  - Enhance error handling with console.error logging
- bmMaterialsController.js: Add response validation check for debugging.

Aditya-gam committed:
refactor(bm): remove debugging logs and unnecessary exception logging

Clean up debugging code and improve logging practices:

- Remove console.error debugging statement from aggregateMaterialCost
- Remove logger.logException calls from validation error handlers
  (handleDateRangeError, handleQueryParamError) - these are expected
  validation errors that return proper HTTP responses, not exceptions
- Remove unused debugging code block (costDataTotal calculation)

Aditya-gam committed:
refactor(bm): remove debugging logs and unnecessary exception logging

Clean up debugging code and improve logging practices:

- Remove console.error debugging statement from aggregateMaterialCost
- Remove logger.logException calls from validation error handlers
  (handleDateRangeError, handleQueryParamError) - these are expected
  validation errors that return proper HTTP responses, not exceptions
- Remove unused debugging code block (costDataTotal calculation)

Aditya-gam committed:
test(bm): fix and update tests for material cost correlation

- Fix mock response chaining (json returns this for proper flow)
- Change mockRejectedValueOnce to mockImplementationOnce for sync functions
  (parseAndNormalizeDateRangeUTC is synchronous, not async)
- Update tests to reflect that validation errors are no longer logged
  as exceptions (query param errors, date errors)
- Fix pipeline stage assertions in helpers test for updated aggregation
  pipeline structure (Stage 3.5/3.6 for unitPrice handling)
- Update error logging assertion to match correct function name
  (buildLookupMaps instead of buildCostCorrelationResponse)
- Remove unused commented-out import (MongoMemoryServer)

Aditya-gam committed:
test(bm): fix and update tests for material cost correlation

- Fix mock response chaining (json returns this for proper flow)
- Change mockRejectedValueOnce to mockImplementationOnce for sync functions
  (parseAndNormalizeDateRangeUTC is synchronous, not async)
- Update tests to reflect that validation errors are no longer logged
  as exceptions (query param errors, date errors)
- Fix pipeline stage assertions in helpers test for updated aggregation
  pipeline structure (Stage 3.5/3.6 for unitPrice handling)
- Update error logging assertion to match correct function name
  (buildLookupMaps instead of buildCostCorrelationResponse)
- Remove unused commented-out import (MongoMemoryServer)

Aditya-gam committed:
Merge branch 'development' into Aditya-feat/material-cost-correlation-chart

Anusha-Gali reviewed (changes requested):
Hi Aditya,

I have reviewed your PR locally and the API response is generated as per requirement. However the npm test has failed. Do have a look into it.

<img width="1273" height="763" alt="Image" src="https://github.com/user-attachments/assets/f4c7a14b-cf6b-48e9-b368-dd7b37ca2d89" />
<img width="1273" height="763" alt="Image" src="https://github.com/user-attachments/assets/3bd64bf9-f8ee-4edf-98dc-feb81cb0c2d8" />
<img width="931" height="763" alt="Image" src="https://github.com/user-attachments/assets/66980f97-f4b2-424a-a6f8-f5cc976bea8e" />

Aditya-gam committed:
fix: fixed the mock of the controller

Aditya-gam committed:
fix: fixed the mock of the controller

Aditya-gam commented:
> Hi Aditya,
> 
> I have reviewed your PR locally and the API response is generated as per requirement. However the npm test has failed. Do have a look into it.
> 
> <img alt="Image" width="1273" height="763" src="https://private-user-images.githubusercontent.com/143547248/547459038-f4c7a14b-cf6b-48e9-b368-dd7b37ca2d89.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzEzNjk4MDIsIm5iZiI6MTc3MTM2OTUwMiwicGF0aCI6Ii8xNDM1NDcyNDgvNTQ3NDU5MDM4LWY0YzdhMTRiLWNmNmItNDhlOS1iMzY4LWRkN2IzN2NhMmQ4OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMjE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDIxN1QyMzA1MDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jODRiYmMwZjFlYjBkNDNiMjMzMmZlZTlkNTNiY2EwOTE1ZTkzNjFjOGY1YTQxYmVjMzlkZmUwY2VkZDQ1MWE3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.mKvlQJW2mH2kxMznfX80OxQaphQZiVbmiLEUepUWsx8"> <img alt="Image" width="1273" height="763" src="https://private-user-images.githubusercontent.com/143547248/547459036-3bd64bf9-f8ee-4edf-98dc-feb81cb0c2d8.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzEzNjk4MDIsIm5iZiI6MTc3MTM2OTUwMiwicGF0aCI6Ii8xNDM1NDcyNDgvNTQ3NDU5MDM2LTNiZDY0YmY5LWY4ZWUtNGVkZi05OGRjLWZlYjgxY2IwYzJkOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMjE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDIxN1QyMzA1MDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01OTJlNWEzMDE1MThkM2MwODJlMGZkZTk3MDEzZDQ5ZDk2MTViNTUwZTdjYjZiMTY1ZmYyYThlOTUyNDk0YTE3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.FQAtl6UbZ7XK2SxBn9ZkKtKDtPGwbftFxuc1Tu-2HKo"> <img alt="Image" width="931" height="763" src="https://private-user-images.githubusercontent.com/143547248/547459037-66980f97-f4b2-424a-a6f8-f5cc976bea8e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzEzNjk4MDIsIm5iZiI6MTc3MTM2OTUwMiwicGF0aCI6Ii8xNDM1NDcyNDgvNTQ3NDU5MDM3LTY2OTgwZjk3LWY0YjItNDI0YS1hNmY4LWY1Y2M5NzZiZWE4ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMjE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDIxN1QyMzA1MDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xMTZiZmQwOTBmZDk5Y2Q4YTM5OTBiODYwNjNiNjM4NDNhYjllYjA5MTFiYmQ4NDRhMGQ4MDcxNTAxNTMwNDQ5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.8r-cIc8uKX_AUT98EpLqVVSSwuggMZn_e5kk0c9unp8">

I fixed the issue with the tests. Mock for the controller was not properly configured. Now the tests work.
<img width="1081" height="290" alt="Tests Passing" src="https://github.com/user-attachments/assets/a2c159c2-f3b4-4d68-b893-0e3123d49880" />

Anusha-Gali reviewed (approved):
Hi Aditya,

Thanks for fixing the backend unit test.
<img width="1278" height="762" alt="Image" src="https://github.com/user-attachments/assets/23417497-3418-40d2-8e77-a3f467498217" />

Aditya-gam committed:
Merge branch 'development' into Aditya-feat/material-cost-correlation-chart

Aditya-gam committed:
refactor(helpers): DRY and lint fixes in material cost correlation helpers

- Extract resolveNamesToIds() to unify project/material name resolution
- Use buildBaseMatchForMaterials in aggregateMaterialCost
- Add getOrCreateMergedEntry for mergeUsageAndCostData
- Throw Error instances for NAME_RESOLUTION_ERROR (lint)
- Harden objectIdToString; use optional chaining and Number.parseFloat

Made-with: Cursor

Aditya-gam committed:
refactor(helpers): DRY and lint fixes in material cost correlation helpers

- Extract resolveNamesToIds() to unify project/material name resolution
- Use buildBaseMatchForMaterials in aggregateMaterialCost
- Add getOrCreateMergedEntry for mergeUsageAndCostData
- Throw Error instances for NAME_RESOLUTION_ERROR (lint)
- Harden objectIdToString; use optional chaining and Number.parseFloat

Made-with: Cursor

Aditya-gam committed:
test(helpers): DRY material cost correlation test utilities

- Add expectDateParseError and expectEndOfDayUTC in date utils tests
- Add DEFAULT_DATE_RANGE, DEFAULT_FILTERS, defaultRequestParams, defaultModels in helpers tests

Made-with: Cursor

Aditya-gam committed:
test(helpers): DRY material cost correlation test utilities

- Add expectDateParseError and expectEndOfDayUTC in date utils tests
- Add DEFAULT_DATE_RANGE, DEFAULT_FILTERS, defaultRequestParams, defaultModels in helpers tests

Made-with: Cursor

Aditya-gam committed:
fix(security): validate ObjectIds in queries; refactor BM materials controller

- Use validated ObjectIds in findOne/create/updateOne/findOneAndUpdate (NoSQL injection)
- Validate purchaseId and material._id in bulk update
- Extract validatePurchaseMaterialsBody and performMaterialPurchase (line limit)
- Extract stock-out-risk helpers to reduce complexity
- Use for-of in bmPostMaterialUpdateBulk; Number.parseFloat where applicable

Made-with: Cursor

Aditya-gam committed:
fix(security): validate ObjectIds in queries; refactor BM materials controller

- Use validated ObjectIds in findOne/create/updateOne/findOneAndUpdate (NoSQL injection)
- Validate purchaseId and material._id in bulk update
- Extract validatePurchaseMaterialsBody and performMaterialPurchase (line limit)
- Extract stock-out-risk helpers to reduce complexity
- Use for-of in bmPostMaterialUpdateBulk; Number.parseFloat where applicable

Made-with: Cursor

Aditya-gam committed:
test(router): disable x-powered-by in BM materials router tests

Address SonarQube security hotspot on version disclosure by disabling
x-powered-by on Express app instances used in tests.

Made-with: Cursor

Aditya-gam committed:
test(router): disable x-powered-by in BM materials router tests

Address SonarQube security hotspot on version disclosure by disabling
x-powered-by on Express app instances used in tests.

Made-with: Cursor

Aditya-gam committed:
test(controller): update BM materials tests for ObjectId-safe queries

- Expect ObjectId types in findOne/create for bmPurchaseMaterials
- Add test for 400 when purchaseId is invalid in bmupdatePurchaseStatus
- Use valid ObjectIds in purchase status tests; add mongoose mock isValid
- Introduce DEFAULT_DATE_QUERY in cost-correlation tests for DRY

Made-with: Cursor

Aditya-gam committed:
test(controller): update BM materials tests for ObjectId-safe queries

- Expect ObjectId types in findOne/create for bmPurchaseMaterials
- Add test for 400 when purchaseId is invalid in bmupdatePurchaseStatus
- Use valid ObjectIds in purchase status tests; add mongoose mock isValid
- Introduce DEFAULT_DATE_QUERY in cost-correlation tests for DRY

Made-with: Cursor

Aditya-gam committed:
fix(security): validate material._id in bmPostMaterialUpdateRecord

Validate material._id before use and pass a constructed ObjectId to
updateOne to address SonarQube jssecurity:S5147 (NoSQL injection).
Return 400 for missing or invalid material._id. Update tests to use
valid ObjectIds and add case for invalid material._id.

Made-with: Cursor

Aditya-gam committed:
fix(security): validate material._id in bmPostMaterialUpdateRecord

Validate material._id before use and pass a constructed ObjectId to
updateOne to address SonarQube jssecurity:S5147 (NoSQL injection).
Return 400 for missing or invalid material._id. Update tests to use
valid ObjectIds and add case for invalid material._id.

Made-with: Cursor

rithika-paii committed:
fix: remove duplicate bmGetMaterialStockOutRisk declaration

DeepighaJ reviewed (approved):
Checkout the current branch and run PR locally. The given endpoints are validated and the response was received as expected. The Unit test also pass as expected.
<img width="1600" height="852" alt="Image" src="https://github.com/user-attachments/assets/59e168be-5fc1-4d13-86b2-41926ba1286c" />
<img width="1600" height="852" alt="Image" src="https://github.com/user-attachments/assets/9a2bd801-cbb7-4332-ae53-63b02abf7821" />
<img width="767" height="467" alt="Image" src="https://github.com/user-attachments/assets/ca300766-f7d6-4044-85ec-ee5b470cf1c8" />

kzou55 reviewed (approved):
Ran and tested the following branch locally. 

- Refer here for the frontend [issues](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/4587#pullrequestreview-4584125581)

Verified the following: 
- Unit Test Passes
  <img width="1125" height="313" alt="Image" src="https://github.com/user-attachments/assets/48f3d5a4-9797-4e03-94f1-3fc2cd30f71a" />
- Endpoints results are expected
  <img width="1422" height="966" alt="Image" src="https://github.com/user-attachments/assets/786894cd-6da0-48aa-a6dc-75e985f89697" />

  <img width="1427" height="970" alt="Image" src="https://github.com/user-attachments/assets/db676662-bb77-4991-97eb-00a0764f0730" />

rithika-paii committed:
fix: improve fallback labels for unresolved project/material type references in cost correlation

HemanthNidamanuru reviewed (approved):
Hi,

I tested this backend PR locally.

Both API endpoints are working as expected, and the related test files are also passing successfully.

<img width="695" height="357" alt="Image" src="https://github.com/user-attachments/assets/0961d605-69c4-4640-b92d-c27ea1990650" />

<img width="1783" height="1077" alt="Image" src="https://github.com/user-attachments/assets/524852f5-9ea3-4942-b88e-c500cbfa1a08" />

<img width="1742" height="1051" alt="Image" src="https://github.com/user-attachments/assets/6d28b03d-0b8b-42b1-9999-d9218aaf41f7" />

kzou55 reviewed (approved):
Hi Rithika, 

I re-reviewed the PR. 

Can confirm everything still is functional. 

## Testing

- Unit Testing

<img width="1256" height="322" alt="Image" src="https://github.com/user-attachments/assets/3b4db2aa-4ed1-47b3-8c0e-4085b15db9f5" />

- Endpoints

    
    <img width="1430" height="949" alt="Image" src="https://github.com/user-attachments/assets/c19d503c-2370-4270-9a6d-e0cb4fc667a8" />

    <img width="1427" height="946" alt="Image" src="https://github.com/user-attachments/assets/0f4dcd6b-a6d3-49ec-bc32-d6515d2de19c" />

BosuBose132 reviewed (approved):
Tested PR #1957 locally. I checked the new Material Cost Correlation backend endpoint with no filters, project filter, material filter, date range filter, and combined filters. The endpoint returned the expected meta and data structure, including project details, quantity used, total cost, totalCostK, costPerUnit, and material type breakdowns. I also tested invalid date and unmatched filter cases to confirm the API handled errors without crashing. The endpoint worked as expected.

<img width="564" height="451" alt="Image" src="https://github.com/user-attachments/assets/cbeb6dbc-adbb-42cf-8f5a-03e9a255ad2d" />
<img width="572" height="497" alt="Image" src="https://github.com/user-attachments/assets/a4bee50f-1987-44ed-971b-1fef89325085" />

rithika-paii committed:
fix: resolve SonarQube backend issues in material cost correlation

rithika-paii committed:
fix: correct objectIdToString to handle non-ObjectId id-like objects

rithika-paii committed:
test(controller): raise bmMaterialsController coverage on new code

rithika-paii committed:
fix: use meaningful toString check for objectIdToString to avoid default stringification

DeepighaJ reviewed (approved):
Tested the PR locally in postman.
The endpoints works as expected, the api returns the expected response for filtered dates which matches the UI.
<img width="1307" height="762" alt="Image" src="https://github.com/user-attachments/assets/f6e70f68-bab8-45a0-878c-5169f3f5e221" />
<img width="517" height="877" alt="Image" src="https://github.com/user-attachments/assets/565e8399-0a55-439c-815f-19d6388fbb32" />
<img width="1306" height="745" alt="Image" src="https://github.com/user-attachments/assets/864bea57-d20b-4df8-8224-e2f9322e0787" />

iAbhi001 reviewed (approved):
Approved, Everything works as expected. The API returns whatever is required.

<img width="1280" height="832" alt="Image" src="https://github.com/user-attachments/assets/85c87454-774e-4aee-b355-e7ea1181e1d0" />

<img width="1280" height="832" alt="Image" src="https://github.com/user-attachments/assets/f86d80d1-12e8-4c14-8646-875ac8998a99" />

sawantgayatri19 reviewed (approved):
Hi team,

I have taken a final look at this PR. All the backend changes for the Material Cost Correlation endpoint, query filters, date normalization utilities, and unit tests are fully implemented, tested, and passing successfully. SonarQube quality gates are also green. 

Everything looks solid and ready to merge!
